libraries/SoftwareSerial/src/SoftwareSerial.cpp

```diff
@@ -0,0 +1,518 @@
+    if (!is_half_duplex()) return;
+
+    if (rx)
+    {
+        tx_active = false;
+        setup_rx();
+
+        rx_bit_count = -1; // waiting for start bit
+        rx_wait_ticks = 2; // wait 2 bit times for start bit
+        rx_active = true;
+    }
+    else
+    {
+        if (rx_active)
+        {
+            rx_active = false;
+            setup_tx();
+        }
+    }
+}
+
+//
+// ISR
+//
+
+void SoftwareSerial::do_rx()
+{
+    // if not enabled, do nothing
+    if (!rx_active) return;
+
+    // if tick count is non-zero, continue waiting
+    rx_wait_ticks--;
+    if (rx_wait_ticks > 0) return;
+
+    // read bit, invert if inverted logic
+    const bool bit = GPIO_GetBit(rx_pin) ^ invert;
+
+    // waiting for start bit?
+    if (rx_bit_count == -1)
+    {
+        // is start bit?
+        // this is UART, so idle line is high and start bit is going low
+        if (!bit)
+        {
+            rx_frame = 0;
+            rx_bit_count = 0;
+
+            // wait 1 1/2 bit times to sample in the middle of the bit
+            rx_wait_ticks = SOFTWARE_SERIAL_OVERSAMPLE + (SOFTWARE_SERIAL_OVERSAMPLE >> 1);
+        }
+        else
+        {
+            // waiting for start bit, but didn't get it
+            // wait for next interrupt to check again
+            rx_wait_ticks = 1;
+        }
+    }
+    else if (rx_bit_count >= 8) // waiting for stop bit?
+    {
+        // is stop bit?
+        // this is UART, so stop bit (== idle line) is high
+        if (bit)
+        {
+            // add byte to buffer
+            bool overflow;
+            rx_buffer->push(rx_frame, true, overflow);
+
+            // avoid overwriting overflow flag
+            if (overflow) did_rx_overflow = true;
+        }
+
+        // assume frame is completed, wait for next start bit at next interrupt
+        // even if there was no stop bit
+        rx_bit_count = -1;
+        rx_wait_ticks = 1;
+    }
+    else // data bits
+    {
+        rx_frame >>= 1;
+        if (bit) rx_frame |= 0x80;
+        rx_bit_count++;
+        rx_wait_ticks = SOFTWARE_SERIAL_OVERSAMPLE;
+    }
+
+}
+
+void SoftwareSerial::do_tx()
+{
+    // if not enabled, do nothing
+    if (!tx_active) return;
+
+    // if tick count is non-zero, continue waiting
+    tx_wait_ticks--;
+    if (tx_wait_ticks > 0) return;
+
+    // all bits in frame sent?
+    if (tx_bit_count >= 10)
+    {
+        // if no frame is pending and half-duplex, switch to RX mode
+        // otherwise, we're done transmitting
+        if (tx_pending || !is_half_duplex())
+        {
+            tx_active = false;
+        }
+        else if (is_half_duplex() && isListening())
+        {
+            // wait HALF_DUPLEX_SWITCH_DELAY bits before switching to RX
+            // NOTE: in STM32dunio, they wait for 10 + (OVERSAMPLE * HALF_DUPLEX_SWITCH_DELAY) bits. i believe this is a bug on their end tho
+            if (tx_bit_count >= 10 + SOFTWARE_SERIAL_HALF_DUPLEX_SWITCH_DELAY)
```

Comment:
in this branch, due to the tx_wait_ticks=1, tx_bit_count increments on every tick. So multiplying the switch delay by OVERSAMPLE is correct, even if unintuitive